From: hmpthcs <homopithecus@protonmail.com>
Date: Sun, 30 Jul 2023 00:59:02 -0400
Subject: rockchip-ebc: add tunable thresholds to mw_mode = 3

add ability to tune low, mid and hi thresholds for bw_mode = 3;
mw_mode = 3 is meant to quantize pixel values into four regions for use
with the DU4 mode.
---
 drivers/gpu/drm/rockchip/rockchip_ebc.c | 25 +++++++++++++++++++------
 1 file changed, 19 insertions(+), 6 deletions(-)

diff --git a/drivers/gpu/drm/rockchip/rockchip_ebc.c b/drivers/gpu/drm/rockchip/rockchip_ebc.c
index c34a5dde..8b217f1 100644
--- a/drivers/gpu/drm/rockchip/rockchip_ebc.c
+++ b/drivers/gpu/drm/rockchip/rockchip_ebc.c
@@ -250,6 +250,7 @@ MODULE_PARM_DESC(delay_c, "delay_c");
 // mode = 0: 16-level gray scale
 // mode = 1: 2-level black&white with dithering enabled
 // mode = 2: 2-level black&white, uses bw_threshold
+// mode = 3: 4-level gray scale
 static int bw_mode = 0;
 module_param(bw_mode, int, S_IRUGO|S_IWUSR);
 MODULE_PARM_DESC(bw_mode, "black & white mode");
@@ -258,6 +259,18 @@ static int bw_threshold = 7;
 module_param(bw_threshold, int, S_IRUGO|S_IWUSR);
 MODULE_PARM_DESC(bw_threshold, "black and white threshold");
 
+static int fourtone_low_threshold = 4;
+module_param(fourtone_low_threshold, int, S_IRUGO|S_IWUSR);
+MODULE_PARM_DESC(fourtone_low_threshold, "everything below this is white");
+
+static int fourtone_mid_threshold = 7;
+module_param(fourtone_mid_threshold, int, S_IRUGO|S_IWUSR);
+MODULE_PARM_DESC(fourtone_mid_threshold, "from low_threshold to here is light gray; from here to hi_threhsold is dark gray");
+
+static int fourtone_hi_threshold = 12;
+module_param(fourtone_hi_threshold, int, S_IRUGO|S_IWUSR);
+MODULE_PARM_DESC(fourtone_hi_threshold, "everything above this is black");
+
 static int bw_dither_invert = 0;
 module_param(bw_dither_invert, int, S_IRUGO|S_IWUSR);
 MODULE_PARM_DESC(bw_dither_invert, "invert dither colors in bw mode");
@@ -2011,21 +2024,21 @@ static bool rockchip_ebc_blit_fb_xrgb8888(const struct rockchip_ebc_ctx *ctx,
 				case 3:
 					// downsample to 4 bw values corresponding to the DU4
 					// transitions: 0, 5, 10, 15
-					if (rgb0 < 4){
+					if (rgb0 < fourtone_low_threshold){
 						rgb0 = 0;
-					} else if (rgb0  < 8){
+					} else if (rgb0  < fourtone_mid_threshold){
 						rgb0 = 5;
-					} else if (rgb0  < 12){
+					} else if (rgb0  < fourtone_hi_threshold){
 						rgb0 = 10;
 					} else {
 						rgb0 = 15;
 					}
 
-					if (rgb1 < 4){
+					if (rgb1 < fourtone_low_threshold){
 						rgb1 = 0;
-					} else if (rgb1  < 8){
+					} else if (rgb1  < fourtone_mid_threshold){
 						rgb1 = 5;
-					} else if (rgb1  < 12){
+					} else if (rgb1  < fourtone_hi_threshold){
 						rgb1 = 10;
 					} else {
 						rgb1 = 15;
